scipy/signal/filter_design.py

```diff
@@ -329,21 +343,47 @@ def freqz(b, a=1, worN=None, whole=False, plot=None):
 
     """
     b, a = map(atleast_1d, (b, a))
-    if whole:
-        lastpoint = 2 * pi
-    else:
-        lastpoint = pi
-    if worN is None:
-        N = 512
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
-    elif isinstance(worN, int):
-        N = worN
-        w = numpy.linspace(0, lastpoint, N, endpoint=False)
+
+    worN = 512 if worN is None else worN  # historically we allowed None
+
+    h = None
+    if isinstance(worN, int):
+        lastpoint = 2 * pi if whole else pi
+        w = numpy.linspace(0, lastpoint, worN, endpoint=False)
+        min_size = max(a.shape[axis], b.shape[axis])
+        if worN >= min_size and fftpack.next_fast_len(worN) == worN:
+            # if worN is fast, 2 * worN will be fast, too,
+            # so no need to check that one
+            n_fft = worN if whole else worN * 2
+            if np.isrealobj(b) and np.isrealobj(a):
+                h = np.fft.rfft(b, n=n_fft, axis=axis)[:worN]
+                if a.size == 1:
+                    h /= a
+                else:
+                    h /= np.fft.rfft(a, n=n_fft, axis=axis)[:worN]
+                if whole:
+                    # exclude DC and maybe Nyquist
+                    stop = h.shape[axis] - 1 if n_fft % 2 == 1 else -2
+                    h_flip = [slice(None)] * h.ndim
+                    h_flip[axis] = slice(stop, 0, -1)
+                    h = np.concatenate((h, h[tuple(h_flip)].conj()), axis=axis)
```

Comment:
Here we need slice manipulation because the `stop` point varies, too -- and if we're going to have that we might as well incorporate the reversal directly